ferry.go

```diff
@@ -378,10 +406,17 @@ func (f *Ferry) WaitUntilBinlogStreamerCatchesUp() {
 //
 // This method will actually not shutdown the BinlogStreamer immediately.
 // You will know that the BinlogStreamer finished when .Run() returns.
-func (f *Ferry) FlushBinlogAndStopStreaming() {
+func (f *Ferry) FlushBinlogAndStopStreaming(ctx context.Context) {
 	if f.WaitUntilReplicaIsCaughtUpToMaster != nil {
 		isReplica, err := CheckDbIsAReplica(f.WaitUntilReplicaIsCaughtUpToMaster.MasterDB)
 		if err != nil {
+			// TODO: this method is usually called in the main thread. panicking
+			// here is different from the style where only Run panics (as it runs
+			// in a separate thread).
+			//
+			// Perhaps it's also a better idea to return errors in all situations
```

Comment:
I think wherever we can provide additional insights to the user in the form of an error is best if we can also ensure we're not going to corrupt anything or cause data loss. `panic`ing is definitely a way to ensure we immediately abort, but with the additions of pause/resume we'll need more control 👍